File: android/game/src/main/kotlin/com/example/janggiai/game/Notation.kt
package com.example.janggiai.game

import com.example.janggiai.game.Board.CHO
import com.example.janggiai.game.Board.COLS
import com.example.janggiai.game.Board.sq
import com.example.janggiai.game.Piece.CANNON
import com.example.janggiai.game.Piece.CHARIOT
import com.example.janggiai.game.Piece.ELEPHANT
import com.example.janggiai.game.Piece.GUARD
import com.example.janggiai.game.Piece.HORSE
import com.example.janggiai.game.Piece.KING
import com.example.janggiai.game.Piece.PAWN

/**
 * Human-readable notation (port of `engine/notation.py`).
 *
 * Squares use the Korean two-digit form `RC`: R = row counted from Han's side 1..9 then 0 for the
 * 10th row, C = file 1..9 from Cho's left.  Cho's king starts on `95`, Han's on `25`.
 * Canonical machine move string: `"95-73"` or `"pass"`.
 */
object Notation {
    fun hanja(p: Int): String {
        val cho = Piece.colorOf(p) == CHO
        return when (Piece.typeOf(p)) {
            KING -> if (cho) "楚" else "漢"
            GUARD -> "士"; CHARIOT -> "車"; CANNON -> "包"; HORSE -> "馬"; ELEPHANT -> "象"
            PAWN -> if (cho) "卒" else "兵"
            else -> "?"
        }
    }

    fun korean(p: Int): String {
        val cho = Piece.colorOf(p) == CHO
        return when (Piece.typeOf(p)) {
            KING -> "궁"; GUARD -> "사"; CHARIOT -> "차"; CANNON -> "포"; HORSE -> "마"; ELEPHANT -> "상"
            PAWN -> if (cho) "졸" else "병"
            else -> "?"
        }
    }

    fun sideName(side: Int): String = if (side == CHO) "cho" else "han"
    fun sideKorean(side: Int): String = if (side == CHO) "초" else "한"

    fun squareToString(s: Int): String {
        val r = Board.rowOf(s); val c = Board.colOf(s)
        return "${(r + 1) % 10}${c + 1}"
    }

    fun parseSquare(text: String): Int {
        val t = text.trim()
        if (t.length != 2 || !t.all { it.isDigit() }) throw IllegalArgumentException("bad square '$text'")
        val r = ((t[0] - '0') - 1 + 10) % 10
        val c = (t[1] - '0') - 1
        if (c !in 0 until COLS) throw IllegalArgumentException("bad square '$text'")
        return sq(r, c)
    }

    fun moveToString(m: Int): String =
        if (m == Move.PASS) "pass" else "${squareToString(Move.from(m))}-${squareToString(Move.to(m))}"

    private val MOVE_RE = Regex("^\\s*(\\d\\d)\\s*(?:-|→|->|>)?\\s*(\\d\\d)\\s*$")

    fun parseMove(text: String): Int {
        val t = text.trim().lowercase()
        if (t in setOf("pass", "한수쉼", "쉼", "p", "--")) return Move.PASS
        val mr = MOVE_RE.find(t) ?: throw IllegalArgumentException("bad move '$text'")
        return Move.encode(parseSquare(mr.groupValues[1]), parseSquare(mr.groupValues[2]))
    }

    /** Pretty form with the moving piece: `馬 92→73` / `車 01x04` / `한수쉼`. */
    fun describe(board: IntArray, m: Int): String {
        if (m == Move.PASS) return "한수쉼"
        val from = Move.from(m); val to = Move.to(m)
        val p = board[from]
        val name = if (p != 0) hanja(p) else "?"
        val sep = if (board[to] != 0) "x" else "→"
        return "$name ${squareToString(from)}$sep${squareToString(to)}"
    }
}

File: engine/notation.py
"""Human-readable notation.

Squares use the Korean two-digit convention ``RC``: R is the row counted
from Han's side 1..9 then 0 for the 10th row, C is the file 1..9 counted
from Cho's left.  Cho's king therefore starts on ``95``, Han's on ``25``.

Canonical machine move string: ``"95-73"`` (or ``"pass"``).
"""
from __future__ import annotations

import re

from .board import (
    CANNON,
    CHARIOT,
    CHO,
    ELEPHANT,
    GUARD,
    HORSE,
    KING,
    PASS_MOVE,
    PAWN,
    encode_move,
    move_from,
    move_to,
    piece_color,
    piece_type,
)
from .geometry import COLS, rc, sq

HANJA = {
    CHO: {KING: "楚", GUARD: "士", CHARIOT: "車", CANNON: "包", HORSE: "馬", ELEPHANT: "象", PAWN: "卒"},
    1: {KING: "漢", GUARD: "士", CHARIOT: "車", CANNON: "包", HORSE: "馬", ELEPHANT: "象", PAWN: "兵"},
}
KOREAN = {
    CHO: {KING: "궁", GUARD: "사", CHARIOT: "차", CANNON: "포", HORSE: "마", ELEPHANT: "상", PAWN: "졸"},
    1: {KING: "궁", GUARD: "사", CHARIOT: "차", CANNON: "포", HORSE: "마", ELEPHANT: "상", PAWN: "병"},
}
ENGLISH = {KING: "king", GUARD: "guard", CHARIOT: "chariot", CANNON: "cannon",
           HORSE: "horse", ELEPHANT: "elephant", PAWN: "pawn"}
SIDE_NAME = {CHO: "cho", 1: "han"}
SIDE_KOREAN = {CHO: "초", 1: "한"}


def piece_hanja(p: int) -> str:
    return HANJA[piece_color(p)][piece_type(p)]


def piece_korean(p: int) -> str:
    return KOREAN[piece_color(p)][piece_type(p)]


def square_to_str(s: int) -> str:
    r, c = rc(s)
    return f"{(r + 1) % 10}{c + 1}"


def str_to_square(text: str) -> int:
    text = text.strip()
    if len(text) != 2 or not text.isdigit():
        raise ValueError(f"bad square {text!r}")
    r = (int(text[0]) - 1) % 10
    c = int(text[1]) - 1
    if not 0 <= c < COLS:
        raise ValueError(f"bad square {text!r}")
    return sq(r, c)


def move_to_str(m: int) -> str:
    """Canonical machine form, e.g. ``95-73`` or ``pass``."""
    if m == PASS_MOVE:
        return "pass"
    return f"{square_to_str(move_from(m))}-{square_to_str(move_to(m))}"


_MOVE_RE = re.compile(r"^\s*(\d\d)\s*(?:-|→|->|>)?\s*(\d\d)\s*$")


def str_to_move(text: str) -> int:
    t = text.strip().lower()
    if t in ("pass", "한수쉼", "쉼", "p", "--"):
        return PASS_MOVE
    m = _MOVE_RE.match(t)
    if not m:
        raise ValueError(f"bad move {text!r}")
    return encode_move(str_to_square(m.group(1)), str_to_square(m.group(2)))


def describe_move(board: list[int], m: int) -> str:
    """Pretty form with the moving piece: ``馬 92→73`` / ``車 01x04`` / ``한수쉼``."""
    if m == PASS_MOVE:
        return "한수쉼"
    frm, to = move_from(m), move_to(m)
    p = board[frm]
    name = piece_hanja(p) if p else "?"
    sep = "x" if board[to] else "→"
    return f"{name} {square_to_str(frm)}{sep}{square_to_str(to)}"
